### PR TITLE
refactor: Make `AsyncResult` a `PromiseLike`

### PR DESCRIPTION
**Pull request by zharinov**
## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/result.ts [hunk 1]
```diff
@@ -310,66 +310,68 @@ export class Result<T, E = Error> {
...
+      | ((value: Result<T, E>) => TResult1 | PromiseLike<TResult1>)
+      | undefined
+      | null
+  ): PromiseLike<TResult1> {
+    return this.asyncResult.then(onfulfilled);
   }
 
   static ok<T>(val: NonNullable<T>): AsyncResult<T, never> {
-    return new AsyncResult((resolve) => resolve(Result.ok(val)));
+    return new AsyncResult(new Promise((resolve) => resolve(Result.ok(val))));
```
Comment (viceice) [inline lib/util/result.ts:326]:
```suggestion
    return new AsyncResult(Promise.resolve(Result.ok(val));
```
simplify?

File: lib/util/result.ts [hunk 2]
```diff
@@ -310,66 +310,68 @@ export class Result<T, E = Error> {
...
   }
 
   static err<E>(err: NonNullable<E>): AsyncResult<never, E> {
-    return new AsyncResult((resolve) => resolve(Result.err(err)));
+    return new AsyncResult(new Promise((resolve) => resolve(Result.err(err))));
```
Comment (viceice) [inline lib/util/result.ts:330]: see above

File: lib/util/result.ts [hunk 3]
```diff
@@ -310,66 +310,68 @@ export class Result<T, E = Error> {
...
+            return value;
           }
-
-          return resolve(Result.ok(value));
-        })
-        .catch((err) => resolve(Result.err(err)));
-    });
+          return Result.ok(value);
+        },
+        (err) => {
```
Comment (viceice) [inline lib/util/result.ts:345]: this doesn't catch error from the then block, it only catches errors from existing promise

File: lib/util/result.ts [hunk 4]
```diff
@@ -310,66 +309,68 @@ export class Result<T, E = Error> {
...
   }
 
   static ok<T>(val: NonNullable<T>): AsyncResult<T, never> {
-    return new AsyncResult((resolve) => resolve(Result.ok(val)));
+    return new AsyncResult(Promise.resolve(Result.ok(val)));
   }
 
   static err<E>(err: NonNullable<E>): AsyncResult<never, E> {
-    return new AsyncResult((resolve) => resolve(Result.err(err)));
+    // eslint-disable-next-line promise/no-promise-in-callback
```
Comment (viceice) [inline lib/util/result.ts:329]: 🤔 why and what causes this 🤔

Comment (zharinov) [reply lib/util/result.ts:329]: I think it's bug of linter

Comment (viceice) [reply lib/util/result.ts:329]: we should report it 😉

Comment (zharinov) [reply lib/util/result.ts:329]: If we do, I don't think it well be fixed anytime soon

Comment (zharinov) [reply lib/util/result.ts:329]: But I don't care too much about it

Comment (viceice) [reply lib/util/result.ts:329]: i know, but a bug should be reported. we can leave this comment until upstream fix

Comment (zharinov) [reply lib/util/result.ts:329]: Let's merge it first, I'll see how to provide minimal reproduction

**Review by viceice — APPROVED** [review]

File: lib/util/result.ts (hunk 4 above)
Comment (zharinov) [reply lib/util/result.ts:329]: https://github.com/eslint-community/eslint-plugin-promise/issues/445

**Pull request merged**